### PR TITLE
feat: cleanup ci and patch versions

### LINKED ISSUE #631: [aws] implement --cleanup-ci=.. and support deleting older aliases
support deleting older aliases: e.g. when deploying an alias 10_75_0, just keep last 10_74_X and remove all 10_73_*. similar for deploying v11: just keep v10, remove any 10_X_Y aliases.

### PR DESCRIPTION
**Pull request by dominique-pfister**
fix #631

### CODE REVIEW

File: src/deploy/AWSDeployer.js [hunk 1]
```diff
@@ -546,6 +577,63 @@ export default class AWSDeployer extends BaseDeployer {
...
+  async cleanup() {
+    const { cfg, functionName } = this;
+    const cleanupMatchers = [{
+      name: 'CI',
+      property: 'cleanupCiAge',
+      match: (alias) => alias.match(/^ci(\d+)$/),
+    }, {
+      name: 'patch',
+      property: 'cleanupPatchAge',
+      match: (alias) => alias.match(/^(\d+)_(\d+)_(\d+)(-test)?$/),
```
Comment (dominique-pfister) [inline src/deploy/AWSDeployer.js:589]: tobi and I sometimes use `1_5_7-test`

File: src/deploy/AWSDeployer.js [hunk 2]
```diff
@@ -546,6 +577,63 @@ export default class AWSDeployer extends BaseDeployer {
...
+          Aliases: aliases
+            .filter(({ FunctionVersion }) => version.Version === FunctionVersion)
+            .map(({ Name }) => Name),
+        }));
+
+      for (const { name, property, match } of cleanupMatchers) {
+        if (cfg[property]) {
+          const cleanupBefore = Date.now() - cfg[property];
+          const oldVersions = versions
+            .filter(({ Aliases }) => Aliases.length > 0 && Aliases.every((alias) => match(alias)))
```
Comment (dominique-pfister) [inline src/deploy/AWSDeployer.js:609]: we remove a version only, when *every* alias it has does match, so 2 aliases `ci123` and `ci` pointing to the same version will not be removed

File: src/deploy/AWSDeployer.js [hunk 3]
```diff
@@ -546,6 +577,63 @@ export default class AWSDeployer extends BaseDeployer {
...
+                  FunctionName: functionName,
+                  Name,
+                }));
+              }));
+              await this._lambda.send(new DeleteFunctionCommand({
+                FunctionName: functionName,
+                Qualifier: Version,
+              }));
+              return Version;
+            }, 2);
```
Comment (dominique-pfister) [inline src/deploy/AWSDeployer.js:627]: concurrency greater than 2 causes 429 from AWS

**Comment by dominique-pfister** [comment]
added test in `aws.deployer.test.js` but codecov still complaining (just looks at -g `Integration`?)

File: src/deploy/AWSDeployer.js [hunk 4]
```diff
@@ -604,7 +604,7 @@ export default class AWSDeployer extends BaseDeployer {
 
       for (const { name, property, match } of cleanupMatchers) {
         if (cfg[property]) {
-          const cleanupBefore = Date.now() - cfg[property];
+          const cleanupBefore = Date.now() - (cfg[property] * 1000);
```
Comment (dominique-pfister) [inline src/deploy/AWSDeployer.js:607]: age properties are in seconds, dates in milliseconds

**Review by trieloff — APPROVED** [review]

**Pull request merged**